Our tests run only on UTC, and fail on other TZ's like NZ

Hi,

Looks like our tests were written for UTC only. The following tests fail when running the tests in NZ time zone.

```
isodatetime/tests/test_00.py::TestSuite::test_timepoint_strftime_strptime FAILED
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_now_0 FAILED
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_0 FAILED
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_x FAILED
```

Left the `test_timepoint` running for a few minutes and got no error.

We should handle testing in any time zone. If a test is handling time zones, then these must be parameters to the tests.

Even better, we can have tests running on multiple time zones. That way we would reduce the chances of having issues like #89.

<details>
<summary>
test results
</summary>

```
(venv) kinow@ranma:~/Development/python/workspace/isodatetime$ python setup.py test
running pytest
WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox.
running egg_info
writing metomi_isodatetime.egg-info/PKG-INFO
writing dependency_links to metomi_isodatetime.egg-info/dependency_links.txt
writing entry points to metomi_isodatetime.egg-info/entry_points.txt
writing top-level names to metomi_isodatetime.egg-info/top_level.txt
reading manifest file 'metomi_isodatetime.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'metomi_isodatetime.egg-info/SOURCES.txt'
running build_ext
================================================================ test session starts ================================================================
platform linux -- Python 3.7.6, pytest-5.4.2, py-1.8.1, pluggy-0.13.1 -- /home/kinow/Development/python/workspace/isodatetime/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/kinow/Development/python/workspace/isodatetime, inifile: pytest.ini
plugins: cov-2.9.0, env-0.6.2
collected 55 items                                                                                                                                  

README.md::README.md PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_get_local_time_zone_format PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_get_local_time_zone_no_dst PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_get_local_time_zone_with_dst PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_largest_truncated_property_name PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_smallest_missing_property_name PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timeduration_dumper PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timeduration_parser PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_dump_format PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper_after_copy PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper_bounds_error_message PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper_get_time_zone PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_parser PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_strftime_strptime FAILED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_time_zone PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timerecurrence PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timerecurrence_alt_calendars PASSED
metomi/isodatetime/tests/test_00.py::TestSuite::test_timerecurrence_parser PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_days_in_year_range SKIPPED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_duration_floordiv PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_duration_in_weeks_floordiv PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_duration_subtract PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_duration_to_weeks PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timeduration PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timeduration_add_week PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timepoint_add_duration PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timepoint_add_duration_without_minute PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timepoint_bounds PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timepoint_conflicting_inputs PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timepoint_plus_float_time_duration_day_of_month_type PASSED
metomi/isodatetime/tests/test_01.py::TestDataModel::test_timepoint_subtract PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_diff_time_point_strs PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_format_duration_str_x PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_calendar PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_format PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_now_0 PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_0 PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_1 PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_2 PASSED
metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_x FAILED
metomi/isodatetime/tests/test_exceptions.py::test_isodatetime_error PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_1_bad PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_1_good PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_1_null PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_1_version PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_2_bad PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_2_good PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_3_bad PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_3_good PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_4_bad PASSED
metomi/isodatetime/tests/test_main.py::TestMain::test_4_good PASSED
metomi/isodatetime/tests/test_main.py::test_pipe[2000-args0-2000] PASSED
metomi/isodatetime/tests/test_main.py::test_pipe[2000\n2001-args1-8784.0] PASSED
metomi/isodatetime/tests/test_time_point.py::TestTimePointCompat::test_timepoint SKIPPED

===================================================================== FAILURES ======================================================================
____________________________________________________ TestSuite.test_timepoint_strftime_strptime _____________________________________________________

self = <isodatetime.tests.test_00.TestSuite testMethod=test_timepoint_strftime_strptime>

    def test_timepoint_strftime_strptime(self):
        """Test the strftime/strptime for date/time expressions."""
        parser = parsers.TimePointParser()
        parse_tokens = list(parser_spec.STRFTIME_TRANSLATE_INFO.keys())
        parse_tokens.remove("%z")  # Don't test datetime's tz handling.
        parse_tokens.remove("%j")  # Can't have day_of_year as well as month
        format_string = ""
        for i, token in enumerate(parse_tokens):
            format_string += token
            if i % 2 == 0:
                format_string += " "
            if i % 3 == 0:
                format_string += ":"
            if i % 5 == 0:
                format_string += "?foobar"
            if i % 7 == 0:
                format_string += "++("
        strftime_string = format_string
        strptime_strings = [format_string]
        for key in parser_spec.STRPTIME_EXCLUSIVE_GROUP_INFO.keys():
            strptime_strings[-1] = strptime_strings[-1].replace(key, "")
        strptime_strings.append(format_string)
        for values in parser_spec.STRPTIME_EXCLUSIVE_GROUP_INFO.values():
            for value in values:
                strptime_strings[-1] = strptime_strings[-1].replace(value, "")
        ctrl_date = datetime.datetime(2002, 3, 1, 12, 30, 2)
    
        # Test %z dumping.
        for sign in [1, -1]:
            for hour in range(0, 24):
                for minute in range(0, 59):
                    if hour == 0 and minute == 0 and sign == -1:
                        # -0000, same as +0000, but invalid.
                        continue
                    test_date = data.TimePoint(
                        year=ctrl_date.year,
                        month_of_year=ctrl_date.month,
                        day_of_month=ctrl_date.day,
                        hour_of_day=ctrl_date.hour,
                        minute_of_hour=ctrl_date.minute,
                        second_of_minute=ctrl_date.second,
                        time_zone_hour=sign * hour,
                        time_zone_minute=sign * minute
                    )
                    ctrl_string = "-" if sign == -1 else "+"
                    ctrl_string += "%02d%02d" % (hour, minute)
                    self.assertEqual(test_date.strftime("%z"),
                                     ctrl_string,
                                     "%z for " + str(test_date))
    
        test_date = data.TimePoint(
            year=ctrl_date.year,
            month_of_year=ctrl_date.month,
            day_of_month=ctrl_date.day,
            hour_of_day=ctrl_date.hour,
            minute_of_hour=ctrl_date.minute,
            second_of_minute=ctrl_date.second
        )
        test_date.set_time_zone_to_utc()
        for test_date in [test_date, test_date.copy().to_week_date(),
                          test_date.copy().to_ordinal_date()]:
            ctrl_data = ctrl_date.strftime(strftime_string)
            test_data = test_date.strftime(strftime_string)
>           self.assertEqual(test_data, ctrl_data, strftime_string)
E           AssertionError: '01 :?foobar++(2002-03-0112 03:30 1014985802?foobar02 :12:30:02++(2002 ' != '01 :?foobar++(2002-03-0112 03:30 1014939002?foobar02 :12:30:02++(2002 '
E           - 01 :?foobar++(2002-03-0112 03:30 1014985802?foobar02 :12:30:02++(2002 
E           ?                                       ^^^
E           + 01 :?foobar++(2002-03-0112 03:30 1014939002?foobar02 :12:30:02++(2002 
E           ?                                       ^^^
E            : %d :?foobar++(%F%H %m:%M %s?foobar%S :%X++(%Y

metomi/isodatetime/tests/test_00.py:1259: AssertionError
________________________________________________ TestDateTimeOperator.test_process_time_point_str_x _________________________________________________

self = <isodatetime.tests.test_datetimeoper.TestDateTimeOperator testMethod=test_process_time_point_str_x>

    def test_process_time_point_str_x(self):
        """DateTimeOperator.process_time_point_str(...)
    
        Basic parse and dump of a time point string.
        """
        # 2009-02-13T23:31:30Z
        point_str = str(seconds2point(1234567890))
        datetimeoper = idt_dtoper.DateTimeOperator()
        # Unix time
        self.assertEqual(
            '2019-01-11T10:40:15Z',
            datetimeoper.process_time_point_str(
                'Fri 11 Jan 10:40:15 UTC 2019',
>               print_format=datetimeoper.CURRENT_TIME_DUMP_FORMAT_Z))
E       AssertionError: '2019-01-11T10:40:15Z' != '2019-01-10T22:40:15Z'
E       - 2019-01-11T10:40:15Z
E       ?          ^ ^^
E       + 2019-01-10T22:40:15Z
E       ?          ^ ^^

metomi/isodatetime/tests/test_datetimeoper.py:111: AssertionError
============================================================== short test summary info ==============================================================
SKIPPED [1] metomi/isodatetime/tests/test_01.py:277: need --runslow option to run
SKIPPED [1] metomi/isodatetime/tests/test_time_point.py: need --runslow option to run
FAILED metomi/isodatetime/tests/test_00.py::TestSuite::test_timepoint_strftime_strptime - AssertionError: '01 :?foobar++(2002-03-0112 03:30 101498...
FAILED metomi/isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_x - AssertionError: '2019-01-11T10:40:15Z'...
====================================================== 2 failed, 51 passed, 2 skipped in 4.40s ======================================================
```
</details>

Running `TZ=UTC python setup.py test`, it works.

But the following commands all fail on my laptop.

```
TZ="America/Sao_Paulo" python setup.py test (`AssertionError: '2019-01-11T10:40:15Z' != '2019-01-11T13:40:15Z'`)
TZ="Asia/Colombo" python setup.py test (`AssertionError: '2019-01-11T10:40:15Z' != '2019-01-11T05:10:15Z'`)
TZ="Asia/Tokyo" python setup.py test (`AssertionError: '2019-01-11T10:40:15Z' != '2019-01-11T01:40:15Z'`)
TZ="Pacific/Auckland" python setup.py test (`AssertionError: '2019-01-11T10:40:15Z' != '2019-01-10T22:40:15Z'`)
TZ="NZST-12NZDT,M9.5.0,M4.1.0/3" python setup.py test (`AssertionError: '2019-01-11T10:40:15Z' != '2019-01-10T22:40:15Z'`)
TZ=UTC-12 python setup.py test (`AssertionError: '2019-01-11T10:40:15Z' != '2019-01-10T22:40:15Z'`)
```

Maybe some tests need to have the env TZ enforced to UTC (assuming it was intentional). That could make the tests pass, regardless of the system time zone.

--- a/.travis.yml
+++ b/.travis.yml
@@ -13,10 +13,12 @@ matrix:
     - python: 3.6
       env: RUN_COVERAGE=false
     - python: 3.7
+      env: RUN_COVERAGE=false
+    - python: 3.8
       env: RUN_COVERAGE=true
-    - python: 3.7
-      env: RUN_COVERAGE=false TZ=UTC
-    - python: 3.8-dev
+    - python: 3.8
+      env: RUN_COVERAGE=false TZ=XXX-05:30 # Test in +05:30 time zone
+    - python: 3.9-dev
       env: RUN_COVERAGE=false
     - os: osx
       env: RUN_COVERAGE=false

--- a/metomi/isodatetime/data.py
+++ b/metomi/isodatetime/data.py
@@ -2297,16 +2297,24 @@ def get_timepoint_for_now():
     return get_timepoint_from_seconds_since_unix_epoch(time.time())
 
 
-def get_timepoint_from_seconds_since_unix_epoch(num_seconds):
+def get_timepoint_from_seconds_since_unix_epoch(num_seconds, utc=False):
     """Return a TimePoint at a date/time specified in Unix time.
+
+    Args:
+        num_seconds - The number of seconds since the Unix epoch
+        utc (bool) - Whether the returned TimePoint should be in UTC or the
+            local time zone
 
     Note that Unix time always counts 1 day = 86400 seconds, so if
     we implement leap seconds we need to make the distinction.
-
     """
     reference_timepoint = TimePoint(
         **CALENDAR.UNIX_EPOCH_DATE_TIME_REFERENCE_PROPERTIES)
     reference_timepoint.set_time_zone_to_local()
+    if utc:
+        adj_secs = reference_timepoint.time_zone.get_seconds()
+        reference_timepoint.set_time_zone_to_utc()
+        reference_timepoint += Duration(seconds=adj_secs)
     return reference_timepoint + Duration(seconds=float(num_seconds))
 
 

--- a/metomi/isodatetime/datetimeoper.py
+++ b/metomi/isodatetime/datetimeoper.py
@@ -82,7 +82,6 @@ class DateTimeOperator(object):
                          If not specified, operations use current date time.
 
         """
-        self.parse_formats = self.PARSE_FORMATS
         self.custom_parse_format = parse_format
         self.utc_mode = utc_mode
         if self.utc_mode:
@@ -140,7 +139,7 @@ class DateTimeOperator(object):
             time_point = self.strptime(time_point_str, parse_format)
         else:
             time_point = None
-            for parse_format in self.parse_formats:
+            for parse_format in self.PARSE_FORMATS:
                 try:
                     time_point = self.strptime(time_point_str, parse_format)
                     break
@@ -266,12 +265,12 @@ class DateTimeOperator(object):
     def get_datetime_strptime(self, time_point_str, parse_format):
         """Use the datetime library's strptime as a fallback."""
         point = seconds2point(
-            time.mktime(time.strptime(time_point_str, parse_format)))
-        # FIXME: Neither time.strptime nor datetime.datetime.strptime
-        # returns the value of the %Z (time zone) field, so we ignore that for
-        # now except explicit UTC/GMT in the string.
-        if any(s in time_point_str for s in ('UTC', 'GMT')):
-            point.set_time_zone_to_utc()
+            time.mktime(time.strptime(time_point_str, parse_format)),
+            utc=True)
+        # NOTE: Neither time.strptime nor datetime.datetime.strptime can
+        # cope with %Z (time zone) that isn't 'UTC' or 'GMT'. There is no
+        # way to differentiate between an unrecognised zone (e.g. CET) and
+        # a bad time_point_str
         return point
 
     def process_time_point_str(

--- a/metomi/isodatetime/tests/test_00.py
+++ b/metomi/isodatetime/tests/test_00.py
@@ -1195,32 +1195,60 @@ class TestSuite(unittest.TestCase):
 
     def test_timepoint_strftime_strptime(self):
         """Test the strftime/strptime for date/time expressions."""
-        parser = parsers.TimePointParser()
-        parse_tokens = list(parser_spec.STRFTIME_TRANSLATE_INFO.keys())
-        parse_tokens.remove("%z")  # Don't test datetime's tz handling.
-        parse_tokens.remove("%j")  # Can't have day_of_year as well as month
-        format_string = ""
-        for i, token in enumerate(parse_tokens):
-            format_string += token
-            if i % 2 == 0:
-                format_string += " "
-            if i % 3 == 0:
-                format_string += ":"
-            if i % 5 == 0:
-                format_string += "?foobar"
-            if i % 7 == 0:
-                format_string += "++("
-        strftime_string = format_string
-        strptime_strings = [format_string]
-        for key in parser_spec.STRPTIME_EXCLUSIVE_GROUP_INFO.keys():
-            strptime_strings[-1] = strptime_strings[-1].replace(key, "")
-        strptime_strings.append(format_string)
-        for values in parser_spec.STRPTIME_EXCLUSIVE_GROUP_INFO.values():
-            for value in values:
-                strptime_strings[-1] = strptime_strings[-1].replace(value, "")
-        ctrl_date = datetime.datetime(2002, 3, 1, 12, 30, 2)
+        parser = parsers.TimePointParser(assumed_time_zone=(0, 0))
+        strftime_string = "%d :?foobar++(%F%H %j:%m %M?foobar :%S++(%X %Y:"
+        strptime_strings = [
+            "%d :?foo++(%H :%m %M?foo :%S++( %Y:",
+            "%j :?foo %H++( %M?foo :%S++( %Y:",
+            "?foo%s :++("
+        ]
+        ctrl_date = datetime.datetime(2002, 3, 1, 12, 30, 2,
+                                      tzinfo=datetime.timezone.utc)
+        test_date = data.TimePoint(
+            year=ctrl_date.year,
+            month_of_year=ctrl_date.month,
+            day_of_month=ctrl_date.day,
+            hour_of_day=ctrl_date.hour,
+            minute_of_hour=ctrl_date.minute,
+            second_of_minute=ctrl_date.second
+        )
+        test_date.set_time_zone_to_utc()
 
-        # Test %z dumping.
+        for test_date in [test_date, test_date.copy().to_week_date(),
+                          test_date.copy().to_ordinal_date()]:
+            # Test strftime (dumping):
+            ctrl_data = ctrl_date.strftime(strftime_string)
+            test_data = test_date.strftime(strftime_string)
+            self.assertEqual(test_data, ctrl_data, strftime_string)
+
+            # Test strptime (parsing):
+            for strptime_string in strptime_strings:
+                # %s not really supported by datetime
+                if "%s" in strptime_string:
+                    unix_time = ctrl_date.timestamp()
+                    # The `%` below is the string format operator (not modulo)
+                    ctrl_dump = strptime_string % int(unix_time)
+                    ctrl_data = ctrl_date
+                else:
+                    ctrl_dump = ctrl_date.strftime(strptime_string)
+                    ctrl_data = datetime.datetime.strptime(
+                        ctrl_dump, strptime_string)
+                test_dump = test_date.strftime(strptime_string)
+                test_data = parser.strptime(test_dump, strptime_string)
+                test_data.set_time_zone_to_utc()
+
+                self.assertEqual(test_dump, ctrl_dump, strptime_string)
+
+                ctrl_data = (
+                    ctrl_data.year, ctrl_data.month, ctrl_data.day,
+                    ctrl_data.hour, ctrl_data.minute, ctrl_data.second
+                )
+                test_data = tuple(list(test_data.get_calendar_date()) +
+                                  list(test_data.get_hour_minute_second()))
+                self.assertEqual(test_data, ctrl_data,
+                                 test_dump + "\n" + strptime_string)
+
+        # Test %z strftime (dumping):
         for sign in [1, -1]:
             for hour in range(0, 24):
                 for minute in range(0, 59):
@@ -1242,49 +1270,6 @@ class TestSuite(unittest.TestCase):
                     self.assertEqual(test_date.strftime("%z"),
                                      ctrl_string,
                                      "%z for " + str(test_date))
-
-        test_date = data.TimePoint(
-            year=ctrl_date.year,
-            month_of_year=ctrl_date.month,
-            day_of_month=ctrl_date.day,
-            hour_of_day=ctrl_date.hour,
-            minute_of_hour=ctrl_date.minute,
-            second_of_minute=ctrl_date.second
-        )
-        test_date.set_time_zone_to_utc()
-        for test_date in [test_date, test_date.copy().to_week_date(),
-                          test_date.copy().to_ordinal_date()]:
-            ctrl_data = ctrl_date.strftime(strftime_string)
-            test_data = test_date.strftime(strftime_string)
-            self.assertEqual(test_data, ctrl_data, strftime_string)
-            for strptime_string in strptime_strings:
-                ctrl_dump = ctrl_date.strftime(strptime_string)
-                test_dump = test_date.strftime(strptime_string)
-                self.assertEqual(test_dump, ctrl_dump, strptime_string)
-                if "%s" in strptime_string:
-                    # The datetime library can't handle this for strptime!
-                    ctrl_data = ctrl_date
-                else:
-                    ctrl_data = datetime.datetime.strptime(
-                        ctrl_dump, strptime_string)
-                test_data = parser.strptime(test_dump, strptime_string)
-                if any(s in strptime_string for s in ('%s', '%z')):
-                    test_data.set_time_zone_to_utc()
-
-                ctrl_data = (
-                    ctrl_data.year, ctrl_data.month, ctrl_data.day,
-                    ctrl_data.hour, ctrl_data.minute, ctrl_data.second
-                )
-                test_data = tuple(list(test_data.get_calendar_date()) +
-                                  list(test_data.get_hour_minute_second()))
-                if "%y" in strptime_string:
-                    # %y is the decadal year (00 to 99) within a century.
-                    # The datetime library, for some reason, sets a default
-                    # century of '2000' - so nuke this extra information.
-                    ctrl_data = tuple([ctrl_data[0] % 100] +
-                                      list(ctrl_data[1:]))
-                self.assertEqual(test_data, ctrl_data, test_dump + "\n" +
-                                 strptime_string)
 
     def test_timerecurrence_alt_calendars(self):
         """Test recurring date/time series for alternate calendars."""

